src/openzaak/components/zaken/api/validators.py

```diff
@@ -280,6 +281,94 @@ def __call__(self, attrs: dict):
                 raise serializers.ValidationError(self.message, code=self.code)
 
 
+class DeelzaakReopenValidator:
+    """
+    Validate that the hoofdzaak is not closed  when a deelzaak is reopened.
+    """
+
+    code = "hoofdzaak-closed"
+    message = (
+        "De hoofdzaak is afgesloten."
+        "Een deelzaak kan pas worden geopend als de hoofzaak dat al is."
+    )
+
+    def __call__(self, attrs: dict):
+        if attrs.get("__is_eindstatus"):
+            return
+
+        zaak = attrs.get("zaak")
+        if not zaak or not zaak.hoofdzaak or not zaak.hoofdzaak.current_status:
+            return
+
+        if zaak.hoofdzaak.current_status.statustype.is_eindstatus():
+            raise serializers.ValidationError(self.message, code=self.code)
+
+
+class EndStatusDeelZakenValidator:
+    """
+    Validate that related deelzaken are closed when the end status is
+    being set.
+
+    The serializer sets the __is_eindstatus attribute in the data dict as
+    part of the ``to_internal_value`` method.
+    """
+
+    code = "deelzaken-closed"
+    message = (
+        "Er zijn gerelateerde deelzaken die nog niet zijn afgesloten zijn."
+        "Deze deelzaken moeten eerst afgesloten worden voordat de zaak afgesloten kan worden."
+    )
+
+    def __call__(self, attrs: dict):
+        if not attrs.get("__is_eindstatus"):
+            return
+
+        zaak = attrs.get("zaak")
+        # earlier validation failed possibly
+        if not zaak:
+            return
+
+        self.check_status_exists(zaak.deelzaken)
+        self.check_internal_status_types(zaak.deelzaken.filter(_zaaktype__isnull=False))
+        self.check_external_status_types(
+            zaak.deelzaken.filter(_zaaktype_relative_url__isnull=False)
+        )
+
+    def check_status_exists(self, qs):
+        if qs.filter(status__isnull=True).exists():
+            raise serializers.ValidationError(self.message, code=self.code)
+
+    def check_internal_status_types(self, qs):
+        eind_statustypevolgnummer = (
+            StatusType.objects.filter(zaaktype=OuterRef("_zaaktype"))
+            .order_by("-statustypevolgnummer")
+            .values("statustypevolgnummer")[:1]
+        )
+
+        current_status_volgnummer = (
+            Status.objects.filter(zaak=OuterRef("pk"))
+            .select_related("statustype")
+            .values("_statustype__statustypevolgnummer")[:1]
```

Comment:
it might be good to explicitly add the ordering here, so we're sure it picks the current status